### DISCUSSION TITLE
I want to achieve an effect where, as the mouse moves on the screen, the nearest corresponding point on the model can always be found.

### DISCUSSION THREAD

**Discussion by MYDJH** [iTwin/itwinjs-core#7172] [Q&A] [answered]
Hello, I want to achieve an effect where, as the mouse moves on the screen, the nearest corresponding point on the model can always be found, similar to the effect in the iTwin.js examples.
Here is the example ：link:https://www.itwinjs.org/sandboxes/iTwinPlatform/Closest%20Point%20on%20Curve
Effect image 
![image](https://github.com/user-attachments/assets/5b6440db-01c4-464a-9814-85d2173adcdb)
In my model image, I want to achieve the effect of the green part, similar to the red dashed line.
![image](https://github.com/user-attachments/assets/1572f838-679e-4c9f-9445-3ba9460764c7)
My current approach is to extract the data for this part of the model and then use this data to create objects such as LineSegment3d and Arc3d, and then use the closestPoint method to achieve this effect. However, I found that there is a TransitionSpiral type of curve, which cannot be directly created and does not have a closestPoint method. I am not sure if this approach is correct, and whether there are better ways to achieve this effect.
 This is the data displayed by my model.
![image](https://github.com/user-attachments/assets/3cf8849a-4c6f-4633-a9c1-65025160156b)
Thank you very much for reading. If you have any suggestions, I would be very grateful.

**Comment by bbastings** [comment, marked as answer]
TransitionSpiral is just another sub-class of CurvePrimitive, as is Arc3d, LineSegment3d, etc. Use GeometryStreamIterator to get **all** curve geometry as a CurvePrimitive and call the closestPoint method.

You don't need to create the CurvePrimitive sub-classes like LineSegment3d/Arc3d directly yourself, the iterator already did that for you and returned them as the base geometry class for curves, surfaces, and solids: GeometryQuery.

You can use GeometryQuery.geometryCategory to test if the geometry is represented by a CurvePrimitive. Then use CurvePrimitive.curvePrimitiveType to test for specific sub-classes and use instanceof if you need to want/need to call class specific methods.

If you just need to call closestPoint, you don't need to care whether it's a line, arc, or spiral...

Hopefully I understood your question...

**Reply by MYDJH** [reply]
"Your understanding is correct, and I am very grateful for your help."
